Add support for prototype affine quantization in pt2e flow

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141421

Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Add support for prototype affine quantization in pt2e flow

Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

cccclai commented:
Thank you for putting it up quickly! Let me get some tests for it.

kimishpatel reviewed (dismissed):
WIP review, leaving comments as I review

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1583,6 +1583,242 @@ def calculate_qparams(self):
+    """How floating point number is mapped to integer number
+
+    symmetric mapping means floating point range is symmetrically mapped to integer range
+    let's say we have floating point range (-3.5, 10.2) and integer range (-8, 7) (int4)
```
it's similar for uint4, basically symmetric quant (+ preserve_zero) will first align the floating point range to be from (-X, X) first, in this case (-10.2, 10.2) and then map it to [0, 15]

the granularity is decided by the side with larger absolute value, in this case 10.2

kimishpatel reviewed (commented):
is this diff ready for review btw?

THe other thing I want to discuss is any thoughts @TiRune has on this for supporting hierarchical scales. 

I think that is somewhat novel.

jerryzh168 commented:
> is this diff ready for review btw?
> 
> THe other thing I want to discuss is any thoughts @TiRune has on this for supporting hierarchical scales.
> 
> I think that is somewhat novel.

@kimishpatel yeah it's ready for review, but probably don't need to review the part that's copied from torchao (or do it at a later time), focusing on pt2e flow part would be good enough

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1583,6 +1583,242 @@ def calculate_qparams(self):
+
+
+@dataclass(frozen=True)
+class PerTensor(Granularity):
```
yeah, we can discuss separately if you want to, reason of this being a dataclass instead of single enum is because other types need to take argument, e.g. PerGroup

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1583,6 +1583,242 @@ def calculate_qparams(self):
+
+
+@dataclass(frozen=True)
+class PerGroup(Granularity):
```
this will just be last axis, for more general configs can use `PerBlock` (need to be added)

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1583,6 +1583,242 @@ def calculate_qparams(self):
+    """
+    Represents row-wise granularity in quantization.
+
+    This is a special case of per-axis quantization and is unique to Float8 matmuls
+    where the input is quantized with a block_size of (1, ..., input.shape[-1]). And the weight
+    is quantized with a block_size of (1, weight.shape[1]).
```
it is sufficient, just more of trying to match the existing name used by fp8..

jerryzh168 commented:
> THe other thing I want to discuss is any thoughts @TiRune has on this for supporting hierarchical scales.
> 
> I think that is somewhat novel.

yeah hierachical scale will require different observer and q/dq ops, but can follow the same flow I think

cccclai reviewed (approved):
LGTM, should be good enough to unblock block-wise quantization with prepare_pt2e

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for prototype affine quantization in pt2e flow"


Summary:
duplicated affine quantization functionality including
observer (https://github.com/pytorch/ao/blob/main/torchao/quantization/observer.py)
and some quant_primitive ops (https://github.com/pytorch/ao/blob/7c3c51fd0de33307e43a1769883a348861d6f7c9/torchao/quantization/quant_primitives.py#L26-L30)
to allow for per group quantization min max observer in pt2e flow

Next: We can follow up to add moving average min max observer

Test Plan:
python test/test_quantization.py -k test_channel_group_quantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge
